doc/design.md

```diff
@@ -0,0 +1,1149 @@
+
+// Add spectrum-to-detector mapping and spectrum numbers.
+std::vector<std::vector<gsl::index>> grouping = {{0, 2}, {1}, {}};
+d.insert<Coord::DetectorGrouping>({Dim::Spectrum, 3}, grouping);
+d.insert<Coord::SpectrumNumber>({Dim::Spectrum, 3}, {1, 2, 3});
+
+// Add time-of-flight axis ("X" in current Mantid nomenclature) which
+// will be shared for all spectra.
+d.insert<Coord::Tof>({Dim::Tof, 1000});
+Dimensions dims({{Dim::Tof, 1000}, {Dim::Spectrum, 3}});
+
+// Add data and uncertainties ("Y" and "E" in current Mantid nomenclature).
+d.insert<Data::Value>("sample", dims);
+d.insert<Data::Variance>("sample", dims);
+
+// Add another set of data and uncertainties.
+d.insert<Data::Value>("background", dims);
+d.insert<Data::Variance>("background", dims);
+
+// Add monitors as a nested dataset.
+d.insert<Coord::MonitorName>({Dim::Monitor, 2}, "beam-status", "main");
+d.insert<Data::Monitor>("monitor", {Dim::Monitor, 2});
+// Add data to monitors. Use DatasetIndex to do lookup by name.
+DatasetIndex<Coord::MonitorName> monitors(d);
+// First monitor has histogram data and is pixelated.
+int pixels = 10 * 10;
+auto &beam_mon = d.get<Data::Monitor>()[monitors["beam-status"]];
+beam_mon.insertAsEdge<Coord::Tof>("", {Dim::Tof, 101});
+beam_mon.insert<Data::Value>("", {{Dim::Tof, 100}, {Dim::Spectrum}},
+                             100 * pixels);
+beam_mon.insert<Data::Variance>("", {{Dim::Tof, 100}, {Dim::Spectrum}},
+                                100 * pixels);
+// Second monitor is event-mode.
+auto &main_mon = d.get<Data::Monitor>()[monitors["main"]];
+// Note that there is Coord::Tof and Data::Tof, the latter is used for events.
+main_mon.insert<Data::Tof>("", {Dim::Event, 238576});
+main_mon.insert<Data::PulseTime>("", {Dim::Event, 238576});
+
+// Convert monitor do histogram.
+auto mon = d.get<Data::Monitor>()[monitors["main"]];
+// Bin events using axis from d.
+auto mon_hist = binEvents(mon, d.at<Coord::Tof>());
+// Merge histogram monitor data into monitor. Note that we can keep the
+// events if we want.
+mon.merge(mon_hist);
+
+// Normalize "sample" and "background" to monitor. Note that the variables
+// in monitor do not have a name so it will affect all data variables with a
+// matching tag.
+d /= d.get<Data::Monitor>()[monitors["main"]];
+
+// Subtract the background. To get around name matching we extract data for
+// subtraction, removing it from the dataset to avoid failure in operator-.
+d.merge(d.extract("sample") - d.extract("background"));
+
+// Copy the dataset. All variables use copy-on-write, so this is cheap.
+auto spinUp(d);
+spinUp.insert<Coord::Polarization>({}, Spin::Up);
+auto spinDown(d);
+spinDown.insert<Coord::Polarization>({}, Spin::Down);
+
+// Build equivalent to Mantid's WorkspaceSingleValue.
+Dataset offset;
+offset.insert<Data::Value>("sample - background", {}, {1.0});
+offset.insert<Data::Variance>("sample - background", {}, {0.1});
+// Note the use of name "sample" such that offset affects sample, not
```

Comment:
I did not understand this comment. Which use of "sample" are you refering to exactly? I can only see "sample - background"